### PR TITLE
Fix #691 by supporting a password manager argument

### LINKED ISSUE #691: HTTP auth with urllib3 does not prompt for credentials
Since 0.19.0, pushing to a repo with HTTP authentication seems not to work, because it fails with HTTP 401 rather than prompting me for credentials. (I am using dulwich mostly via hg-git version 0.8.12.)

With versions prior to 0.19.0, I see something like the following when pushing to one of my repos.

```
$ hg push git+https://github.com/stephenrkell/liballocs.git
pushing to git+https://github.com/stephenrkell/liballocs.git
http authorization required for https://github.com/stephenrkell/liballocs.git/info/refs
realm: GitHub
user: stephenrkell
password: 
```
whereas from 0.19.0 onwards, I get the following.
```
$ hg push git+https://github.com/stephenrkell/liballocs.git
pushing to git+https://github.com/stephenrkell/liballocs.git
abort: git remote error: unexpected http resp 401 for https://github.com/stephenrkell/liballocs.git/info/refs?service=git-receive-pack
```
As an experiment, I backed out 465addec72ee and 0c99ef824c74 (sorry, Mercurial hashes :-) ) and the problem seems to go away.

Let me know if I can help debug further....

### PR DESCRIPTION
**Pull request by byteit101**
If a password manager is provided, retry requests with auth if they fail with 401 errors.
This still requires some changes on hg-git's side to get hg-git working.

I wasn't quite sure how to organize it or what names were best, so let me know if you have suggestions.

`password_mgr` must be an object that has a function of signature `find_user_password(realm, origin) -> (username, password)`. Not sure where the best place to document this is

### CODE REVIEW

File: dulwich/client.py [hunk 1]
```diff
@@ -1588,6 +1592,31 @@ def _http_request(self, url, headers=None, data=None,
...
+            import urllib3.util
+            current_url = urllib3.util.parse_url(resp.geturl())
+            # Create the http origin for the password manager
+            safe_origin = str(urllib3.util.Url(scheme=current_url.scheme,
+                                           host=current_url.host,
+                                           port=current_url.port).url)
+            # Extract the realm (if it exists)
+            realm = None
+            try:
+                realm = resp.headers["WWW-Authenticawte"]
```
Comment (jelmer) [inline dulwich/client.py]: This looks like it can't work

File: dulwich/client.py [hunk 2]
```diff
@@ -1588,6 +1592,31 @@ def _http_request(self, url, headers=None, data=None,
...
+                if realm is not None:
+                    realm = realm[13:-1] # Basic realm=".."
+            except KeyError:
+                pass
+
+            self._username, self._password = \
+                self._password_mgr.find_user_password(realm, safe_origin)
```
Comment (jelmer) [inline dulwich/client.py]: This breaks if find_user_password was not specified.

**Comment by jelmer** [comment]
Would it make more sense to just have HttpGitClient raise an exception when it gets a 401 so the caller can call it again with credentials? That way, the caller can also support other kinds of authentication like negotiate, without having to add support in Dulwich.

This adds quite a bit of untested code, and it will only work for basic auth. The existing (basic-specific) code exists as a compromise because it allows the "password" argument to work for both the HTTP and SSH transports.

File: dulwich/client.py (hunk 1 above)
Comment (byteit101) [reply dulwich/client.py]: Oh whoops, looks like a save key slip.

**Comment by byteit101** [comment]
I created a patch for hg-git that does the same, so I think this PR can probably be closed

https://groups.google.com/forum/#!topic/hg-git/MRzj-CihO9g

**Comment by jelmer** [comment]
Ack, that looks like a much better way of doing this. Thanks, closing this PR.

**Pull request closed without merging**